### PR TITLE
Update ray tune example colab

### PR DESCRIPTION
**Pull request by AyushExel**
Addressed the pointers mentioned [here](https://github.com/wandb/examples/pull/37) 

Edit Summary:
- Added W&B logo
- Added more text explaination
- Improved formatting and spacing
- Removed redundant hyper-parameters

Thanks @charlesfrye, for all the help.

### CODE REVIEW

**Comment by charlesfrye** [comment]
FYI, it looks like you didn't click the checkbox for "Include a link to Colaboratory". We always want to include that so

1. while reviewing, reviewers can open the Colab and see what it looks like (ReviewNB isn't always identical), and
2.  folks coming across our repo can open the notebooks in Colab easily

See the dialog box below:

image

File: colabs/raytune/W&B + RayTune.ipynb [hunk 1]
```diff
@@ -6,7 +6,8 @@
       "name": "W&B-RayTune.ipynb",
```
Comment (charlesfrye) [inline colabs/raytune/W&B + RayTune.ipynb:6]: <p>Drop the <code># Quickstart</code> -- that's just for our Simple Integrations.</p><p> </p><p>You want a space after the <code>#</code> in front of the title, so it looks like:</p><h1>Weights &amp; Biases 💜 Ray/Tune</h1><h1><br></h1><p>Love the text!</p><p> </p><p>I have a few suggested changes (additions in <strong>bold</strong>; subtractions in <s>strikethrough</s>). Watch to make sure you're consistent with <strong>Ray/Tune</strong> rather than <strong>Ray Tune</strong> (with a space).</p><p> </p><blockquote><s>Whereas </s>W&amp;B <strong>(add link)</strong> is a <s>centralized repository for</s> <strong>toolkit with </strong>everything you need to track, reproduce and gain insights from your models easily; Ray/Tune <strong>(add link)</strong> provides a simple interface for scaling and running distributed experiments. 🤝<strong>They're a natural match!🤝</strong></blockquote><blockquote>(<strong>added a newline here</strong>)</blockquote><blockquote><strong>Here's just a</strong> few reasons why our community <strong>💜s </strong>Ray/Tune –</blockquote><blockquote>Simple <s>D</s><strong><s>d</s></strong>istributed execution: Ray<strong>/</strong>Tune makes it easy to scale <strong>all the way </strong>from a single <s>node</s> <strong>laptop</strong><s>, to multiple GPUs, and further multiple nodes.</s> <strong>multiple nodes in a cluster each with multiple GPUs</strong></blockquote><blockquote>Large number of algorithms: Ray<strong>/</strong>Tune has a huge number of algorithms including Population Based Training (<strong>add link)</strong>, ASHA <strong>(add link)</strong>, and HyperBand <strong>(add link)</strong></blockquote><blockquote><s>Framework</s>Method agnostic: Ray<strong>/</strong>Tune works across <strong>deep learning</strong> frameworks <strong>(</strong>including PyTorch, Keras, Tensorflow, <s>XGBoost, and </s>PyTorchLightning<strong>) and with other ML methods like gradient-boosted trees (XGBoost, LightGBM).</strong></blockquote><blockquote>Fault-tolerance: Ray<strong>/</strong>Tune is built on top of Ray, providing fault tolerance out of the box.</blockquote><blockquote><br></blockquote><blockquote><strong>This Colab demonstrates how this integration works for a simple grid search over two hyperparameters. If you've got any questions about the details,</strong></blockquote><blockquote><strong>check out </strong><a href="https://docs.wandb.com/library/integrations/ray-tune" target="_blank"><strong>our documentation</strong></a><strong> or the </strong><a href="https://docs.ray.io/en/master/tune/api_docs/integration.html#weights-and-biases-tune-integration-wandb" target="_blank"><strong>documentation for Ray/Tune</strong></a><strong>.</strong></blockquote><blockquote>Let's start by in<s>a</s>stalling the libraries.</blockquote><p> </p><p>I would suggest bolding the parts in the list before the colons.</p><p> </p><p>Some of this text sounds familiar. Is this copy-pasted from something by Ray? If so, we'll need to paraphrase, as I did with the first line.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (charlesfrye) [inline colabs/raytune/W&B + RayTune.ipynb:6]: <p>I would make this less prominent. Maybe a level 2 or level 3 heading (<code>##</code>  or <code>###</code> )</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (charlesfrye) [inline colabs/raytune/W&B + RayTune.ipynb:6]: <p>There's no need to bold <strong>W&amp;B</strong> at the beginning. <em>Don't forget to always use Ray/Tune</em>.</p><p> </p><p>The text is nice: it makes clear why you'd choose the <code>Logger</code> over the <code>mixin</code> and vice versa. I've got some edits, though.</p><p> </p><p>You open up with "You can either use..." but then end the sentence without saying what else you might use. "Either" should always be paired with "or".</p><p> </p><p>I suggest a rewrite, like so:</p><p> </p><blockquote>1.&nbsp;<code>WandbLogger</code> automatically logs metrics reported to Tune to W&amp;B along with the configuration of the experiment, using <code>tune</code>'s <code>logger</code> interface</blockquote><blockquote>2. The <code>@wandb_mixin</code> decorator gives you greater control over logging by letting you call <code>wandb.log</code> inside the decorated function, allowing you to log custom metrics, plots, and other outputs, like media.</blockquote><blockquote><br></blockquote><blockquote>These methods can be used together or independently. (<strong>no bolding</strong>)</blockquote><blockquote>The example below demonstrates how they can be used together.</blockquote><p>At this point, I would insert another markdown cell, containing the text below, then a code cell with the <code>import</code>s</p><blockquote>First, we import the relevant libraries, including some contents of a useful <code>examples</code>  module from <code>ray.tune</code>. </blockquote><p>then another markdown cell:</p><blockquote> Then, we define our training process, decorated with <code>@wandb_mixin</code>   so we can call <code>wandb.log</code>  to log our custom metric (here, the error rate; you might also log media, e.g. images from the training set).</blockquote><p>and then finally the example code.</p><p> </p><p>Avoid using "Wandb" unless you're referring to an object in code; in that case, it should be in monospaced/"code" style: <code>wandb</code> or <code>WandbOtherStuff</code>. If you're talking about the web app or the company, use "W&amp;B" or "Weights &amp; Biases".</p><p> </p><p>I would recommend that you don't include these examples here, since you're about to demonstrate them below.</p><p>While we're here, though, a few tips: make sure you visually align your code (e.g. <code>"Accuracy"</code> and <code>"Output"</code> should line up). If you're including an ellipsis in Python code (<code>...</code>) you need to make sure that any indentation-changing code is included (here, the <code>return</code>), otherwise it looks a bit weird. The whitespace in this example is also excessive: the lines above, between, and below the keys in the dictionary should be removed.</p><p> </p><p>The <code>WandbLogger</code> example looks weird here, but it's fine in the <a href="https://colab.research.google.com/github/AyushExel/examples/blob/colabs/raytune-torch/colabs/raytune/W%26B%20%2B%20RayTune.ipynb#scrollTo=o2AldspdQLlQ" target="_blank">Colab</a>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (charlesfrye) [inline colabs/raytune/W&B + RayTune.ipynb:6]: <p>Love the comments on Lines 31 + 34!</p><p> </p><p>Lines 1-5: I would import <code>torch.optim</code> and <code>tune</code> at the top: those are "standard" imports. Then, in this cell, import only the special things we need for logging this example: the example itself and the <code>wandb</code> code.</p><p>Line 35: Put spaces between mathematical operators: <code>100 * (1 - acc)</code> . This is part of the<a href="https://www.python.org/dev/peps/pep-0008/" target="_blank"> PEP8 Python style</a>.</p><p>Line 36: There are extra spaces before and after <code>"Error Rate "</code> , which should be <code>"error_rate"</code> (no space at the end!) to match the style of the other metrics (e.g. <code>mean_accuracy</code>). The spacing here is also part of the<a href="https://www.python.org/dev/peps/pep-0008/" target="_blank"> PEP8 Python style</a>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (charlesfrye) [inline colabs/raytune/W&B + RayTune.ipynb:6]: <p>This text is good, but needs a little massaging. Here's my suggested edit:</p><p> </p><blockquote>We're now almost ready to call <code>tune.run</code>  to launch our hyperparameter sweep!</blockquote><blockquote>We just need to do two things:</blockquote><blockquote>1.  Configure the <code>wandb</code>  run with things like a project name</blockquote><blockquote>2.  Pass the <code>WandbLogger</code>  to <code>tune.run</code> so we can capture the output of the <code>tune.report</code> in our training loop.</blockquote><blockquote><br></blockquote><blockquote>The second is easy: we just need to import the <code>WandbLogger</code>  and pass it as one of the <code>loggers</code>  to <code>tune.run</code>.</blockquote><blockquote><br></blockquote><blockquote>You can configure the <code>wandb</code> run inside the <code>config</code> dictionary for your <code>tune</code> run. The W&amp;B configuration info should be stored as a dictionary inside <code>config</code> with the key <code>"wandb"</code>.</blockquote><blockquote><br></blockquote><blockquote>The keys and values in that dictionary are then provided as keyword arguments to <code>wandb.init</code> (see <a href="see [the documentation](https://docs.wandb.com/library/init" target="_blank">the documentation</a> for details), with a few exceptions, described in <a href="https://docs.wandb.com/library/integrations/ray-tune#parameters" target="_blank">our documentation of the Ray/Tune integration</a>.</blockquote><blockquote><br></blockquote><blockquote>The most important of these are <code>api_key</code> and <code>api_key_file</code> , one of which must be provided to allow <code>tune</code> to log information on your behalf to W&amp;B. Here, we're entering it manually using the Python library <code>getpass</code>.</blockquote>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (charlesfrye) [inline colabs/raytune/W&B + RayTune.ipynb:6]: <p>Line 9: This comment is superfluous, I think, because you explain what the <code>Logger</code> does above. Instead, a comment like <code># WandbLogger uses tune.run's logger interface</code> seems like a better choice.</p><p>Line 10: Drop the newline here</p><p>Line 12: Make sure you consistently use either <code>"</code> or <code>'</code> for strings (here, <code>'api_key'</code> ).</p><p>Line 13: Drop the newline here</p><p>Line 16: Drop the newline here</p><p>Line 17: A longer comment would be better: <code># Hyperparameters are set by keyword arguments</code></p><p>Line 20: Watch the alignment of your parens there. Move the <code>}</code> up a line; perhaps move both. This is also part of PEP8.</p><p>Line 25: Why is this included? It's not used or explained.</p><p> </p><p> </p><p>There's another code cell below this one that you should delete.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/40/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

**Comment by charlesfrye** [comment]
Nice work @AyushExel! Looking much better.

Most of my comments were around the text.

Another tranch focused on small bits of code style, whitespace conventions especially. Conforming to PEP8 style is a big part of making Python code more readable by reducing variability. It's the style most Python developers are used to seeing (whether they realize it or not!) and it's a big signal for "this code was written by professionals".

I recommend you look into incorporating PEP8 styling into whatever you use to edit .py files. I use [flake8](https://flake8.pycqa.org/en/latest/), incorporated into vim so that it runs every time I save a file. Eventually, it becomes second nature.

Obviously, PEP8 is both hard to enforce and in some ways (e.g. linewidth compatibility with old terminals) a bad fit for notebooks (and especially Colab!). But the whitespace conventions are helpful and should be followed to the greatest extent possible.

**Comment by charlesfrye** [comment]
I'm bringing this PR in on the colabs branch of the main (wandb) fork of this repo, and then I'm going to implement my ReviewNB suggestions in a commit there.
